File: modules/kwilt-seamless-loop/android/src/main/java/expo/modules/kwiltseamlessloop/KwiltSeamlessLoopModule.kt
package expo.modules.kwiltseamlessloop

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class KwiltSeamlessLoopModule : Module() {
  private val player = LoopAudioTrack()
  private val cache by lazy {
    LoopPcmCache(requireNotNull(appContext.reactContext).applicationContext)
  }

  override fun definition() = ModuleDefinition {
    Name("KwiltSeamlessLoop")
    Events("onStateChanged")

    Function("isAvailable") { true }

    AsyncFunction("prepare") { options: Map<String, Any> ->
      val prepared = cache.prepare(
        uri = options["uri"] as String,
        assetKey = options["assetKey"] as String,
        expectedSampleRateHz = (options["expectedSampleRateHz"] as Number).toInt(),
        expectedChannels = (options["expectedChannels"] as Number).toInt()
      )
      player.prepare(prepared, options["assetKey"] as String).also { sendEvent("onStateChanged", it) }
    }

    AsyncFunction("play") { volume: Double, fadeDurationMs: Double ->
      player.play(volume, fadeDurationMs).also { sendEvent("onStateChanged", it) }
    }
    AsyncFunction("pause") { fadeDurationMs: Double ->
      player.pause(fadeDurationMs).also { sendEvent("onStateChanged", it) }
    }
    AsyncFunction("setVolume") { volume: Double, fadeDurationMs: Double ->
      player.setVolume(volume, fadeDurationMs).also { sendEvent("onStateChanged", it) }
    }
    AsyncFunction("unload") {
      cache.clear()
      player.unload().also { sendEvent("onStateChanged", it) }
    }
    Function("getDiagnostics") { player.diagnostics() }
    AsyncFunction("runContinuityProbe") { loopCount: Int -> unsupportedProbe(loopCount) }
    OnDestroy {
      cache.clear()
      player.unload()
    }
  }

  private fun unsupportedProbe(@Suppress("UNUSED_PARAMETER") loopCount: Int): Map<String, Any?> {
    throw IllegalStateException("Continuity probe is unavailable in this build")
  }
}

File: modules/kwilt-seamless-loop/ios/KwiltSeamlessLoopModule.swift
import AVFAudio
import ExpoModulesCore

struct LoopPrepareOptions: Record {
  @Field var uri: String = ""
  @Field var assetKey: String = ""
  @Field var expectedSampleRateHz: Double = 48_000
  @Field var expectedChannels: Int = 2
}

public final class KwiltSeamlessLoopModule: Module {
  private let cache = LoopPCMCache()
  private let scheduler = LoopScheduler()
  private var notificationTokens: [NSObjectProtocol] = []

  public func definition() -> ModuleDefinition {
    Name("KwiltSeamlessLoop")
    Events("onStateChanged")

    OnCreate {
      self.scheduler.setDiagnosticsHandler { [weak self] diagnostics in
        DispatchQueue.main.async { self?.sendEvent("onStateChanged", diagnostics) }
      }
      self.installAudioNotifications()
    }

    Function("isAvailable") { true }

    AsyncFunction("prepare") { (options: LoopPrepareOptions) async throws -> [String: Any] in
      let prepared = try await self.cache.prepare(
        uri: options.uri,
        assetKey: options.assetKey,
        expectedSampleRateHz: options.expectedSampleRateHz,
        expectedChannels: AVAudioChannelCount(options.expectedChannels)
      )
      return try await self.scheduler.prepare(prepared, assetKey: options.assetKey)
    }

    AsyncFunction("play") { (volume: Double, fadeDurationMs: Double) async throws -> [String: Any] in
      try await self.scheduler.play(volume: volume, fadeDurationMs: fadeDurationMs)
    }

    AsyncFunction("pause") { (fadeDurationMs: Double) async throws -> [String: Any] in
      try await self.scheduler.pause(fadeDurationMs: fadeDurationMs)
    }

    AsyncFunction("setVolume") { (volume: Double, fadeDurationMs: Double) async throws -> [String: Any] in
      try await self.scheduler.setVolume(volume: volume, fadeDurationMs: fadeDurationMs)
    }

    AsyncFunction("unload") { () async -> [String: Any] in
      self.cache.clear()
      return await self.scheduler.unload()
    }

    Function("getDiagnostics") { () -> [String: Any] in
      self.scheduler.getDiagnostics()
    }

    AsyncFunction("runContinuityProbe") { (_: Int) throws -> [String: Any] in
      throw NSError(
        domain: "KwiltSeamlessLoop",
        code: 10,
        userInfo: [NSLocalizedDescriptionKey: "Continuity probe is unavailable in this build"]
      )
    }

    OnDestroy {
      self.removeAudioNotifications()
      self.cache.clear()
      self.scheduler.destroy()
    }
  }

  private func installAudioNotifications() {
    let center = NotificationCenter.default
    notificationTokens.append(center.addObserver(
      forName: AVAudioSession.interruptionNotification,
      object: AVAudioSession.sharedInstance(),
      queue: nil
    ) { [weak self] notification in
      guard let typeValue = notification.userInfo?[AVAudioSessionInterruptionTypeKey] as? UInt,
            let type = AVAudioSession.InterruptionType(rawValue: typeValue) else { return }
      if type == .began {
        self?.scheduler.handleInterruptionBegan()
      } else {
        let optionValue = notification.userInfo?[AVAudioSessionInterruptionOptionKey] as? UInt ?? 0
        self?.scheduler.handleInterruptionEnded(
          shouldResume: AVAudioSession.InterruptionOptions(rawValue: optionValue).contains(.shouldResume)
        )
      }
    })
    notificationTokens.append(center.addObserver(
      forName: AVAudioSession.routeChangeNotification,
      object: AVAudioSession.sharedInstance(),
      queue: nil
    ) { [weak self] _ in self?.scheduler.rebuildAfterConfigurationChange() })
    notificationTokens.append(center.addObserver(
      forName: .AVAudioEngineConfigurationChange,
      object: nil,
      queue: nil
    ) { [weak self] _ in self?.scheduler.rebuildAfterConfigurationChange() })
    notificationTokens.append(center.addObserver(
      forName: AVAudioSession.mediaServicesWereResetNotification,
      object: AVAudioSession.sharedInstance(),
      queue: nil
    ) { [weak self] _ in self?.scheduler.rebuildAfterConfigurationChange() })
  }

  private func removeAudioNotifications() {
    let center = NotificationCenter.default
    notificationTokens.forEach(center.removeObserver)
    notificationTokens.removeAll()
  }
}
